### PR TITLE
perf(multitude): split allocator hot paths from cold refill/oversized…

### PR DESCRIPTION
**Pull request by geeknoid**
… tails

Restructures the hot allocator paths so each public entrypoint inlines to a single bump-fit probe and tail-calls a `#[cold] #[inline(never)]` slow helper for refill and oversized-routing. The `max_normal_alloc` check is deferred from the hot path to the cold tail to avoid a two-level pointer chase through `Arc<ChunkProvider>` on every allocation.

Affected paths:
* `Arena::allocate_layout` / `allocate_shared_layout` (the `Allocator` trait impls used by `bytesbuf` and DST callers) now split into `*_slow` cold tails.
* `try_alloc_slice_local_with` / `_no_drop_with` / `try_alloc_slice_shared_*` gain a `const PANIC: bool` parameter so panicking callers no longer carry a dead `Result<NonNull<_>, _>` niche check through the inlined fast path. Panicking siblings (`alloc_slice_copy`, `alloc_slice_clone_arc`, `alloc_slice_fill_*`, etc.) now call the helpers with `PANIC = true` directly instead of routing through `expect_alloc(try_*)`.
* `alloc_uninit_box` / `_rc` / `_arc` (and the `zeroed_*` variants) fast-path `!needs_drop::<T>` by reusing the simpler value path, skipping the slice helpers per-slot init loop and noop-drop-entry reservation.
* DST `try_reserve_dst_*` paths likewise defer the `max_normal_alloc` comparison into the refill branch.

Other small codegen wins:
* `CurrentChunk::bump_smart_pointers_issued` now hints `n < LARGE - 1` so LLVM can elide overflow guards on the `n + 1` store and subsequent consumers.
* `Vec::resize` / `resize_with` route the length cursor through a `#[mutants::skip] inc_len` helper so `cargo-mutants` cannot mutate `+=` to `*=` (which would freeze the loop counter and turn length-driven tests into infinite-loop timeouts).
* `align_up` is feature-gated to `cfg(feature = "dst", test)` since the non-DST hot paths now reconstruct the aligned pointer inline via `try_bump_fit`.

Tests/tooling:
* Adds `crates/multitude/scripts/correctness.rs`, a self-contained `cargo +nightly -Zscript` runner that reproduces CIs correctness suite locally: Miri (stacked borrows, tree borrows, strict provenance, `-Zmiri-many-seeds=0..32` race coverage), Loom, Bolero property tests, and `cargo-careful`.
* Adds boundary-case tests around `MIN_MAX_NORMAL_ALLOC` and the new fast-path/slow-path split (kills the corresponding `>` → `>=` mutants on the routing decisions).
* Adds `alloc_uninit_*` no-drop fast-path coverage and `allocate_layout_slow_*` mutation-killer tests.

PERF.md regenerated with the post-restructure benchmark numbers.

### CODE REVIEW

File: crates/multitude/src/arena/primitives.rs [hunk 1]
```diff
@@ -92,40 +92,43 @@ impl<A: Allocator + Clone> Arena<A> {
...
+        let align = layout.align().max(1);
+        let bumped = layout.size().max(1);
+
+        // Single bump-fit probe. The `max_normal_alloc` check and the
+        // refill loop are deferred to the cold slow path so the hot
+        // path avoids a 2-level pointer chase through `Arc<ChunkProvider>`.
+        let data_ptr = self.current_local.data_ptr.get();
+        let drop_back_ptr = self.current_local.drop_back.get();
+        let fit = try_bump_fit(data_ptr, drop_back_ptr, align, bumped, 0);
+        if fit.fits {
```
[the remark this answers is not in the document]
Comment (geeknoid) [reply crates/multitude/src/arena/primitives.rs:110]: Good catch — added an explicit `check_isize_overflow(layout.size(), layout.align())?` guard at the top of `allocate_layout`, matching what `allocate_shared_layout` already did. Now layouts with `size + align > isize::MAX` return `Err(AllocError)` before `try_bump_fit`'s `assert_unchecked` can fire on a false predicate.

File: crates/multitude/src/arena/alloc_slice_arc.rs [hunk 2]
```diff
@@ -157,7 +165,9 @@ impl<A: Allocator + Clone> Arena<A> {
         I::IntoIter: ExactSizeIterator,
         A: Send + Sync,
     {
-        expect_alloc(self.try_alloc_slice_fill_iter_arc(iter))
+        let ptr = expect_alloc(self.try_alloc_slice_shared_fill_iter_inner::<_, _, false>(iter));
```
[the remark this answers is not in the document]
Comment (geeknoid) [reply crates/multitude/src/arena/alloc_slice_arc.rs]: Fixed — switched `alloc_slice_fill_iter_arc` to call the inner helper with `PANIC = true` directly, matching the rest of the panicking `alloc_slice_*_arc` siblings. The dead `Result` niche check is now gone from this entrypoint's inlined fast path.

**Review by sandersaares — APPROVED** [review]

**Comment by geeknoid** [comment]
Fixed the CI failures from [run 26268409478](https://github.com/microsoft/oxidizer/actions/runs/26268409478):

**Coverage (`--no-default-features` build):**
- `use crate::internal::constants::MIN_MAX_NORMAL_ALLOC` in `src/arena/tests.rs` was unconditional, but every test using it is `#[cfg(feature = "stats")]`. Gated the import behind `#[cfg(feature = "stats")]`.
- `align_up` in `src/arena/internals.rs` was cfg'd as `any(feature = "dst", test)`, but the in-crate tests that consume it are themselves `#[cfg(feature = "dst")]` — so under `--no-default-features` test build the function became unused. Narrowed the cfg to `feature = "dst"`.

**Windows testing (`bytemuck_integration::try_alloc_{box,rc}_over_aligned_returns_err`):**
- My `!needs_drop::<T>` fast path for `alloc_uninit_box/_rc` / `alloc_zeroed_box/_rc` routed through `try_alloc_inner_value::<MaybeUninit<T>>(MaybeUninit::uninit(), …)`, which passes the value by value. Under coverage instrumentation on Windows that function isn't inlined, so the caller (the test fn) must materialize a 64 KiB-aligned `MaybeUninit<OverAligned>` stack slot — overflowing Windows' default 1 MiB stack (this is exactly the failure mode the existing comment in `bytemuck_integration.rs` warns about for slice/closure paths).
- Reverted the Box and Rc no-drop fast paths to the original closure-based slice path. The closure-based slice path optimizes `dst.write(MaybeUninit::uninit())` to a no-op (writing uninit bytes is a no-op), so no stack materialization happens. Added a comment explaining the constraint so this isn't reintroduced.
- The Arc variants are unaffected: they already use a function-reference initializer (`MaybeUninit::uninit` as `fn()`), not a by-value parameter, so no stack materialization in the caller.

Verified locally:
- `cargo nextest run --no-default-features -p multitude --lib --tests` → 950 / 950 pass
- `cargo nextest run --all-features -p multitude --lib --tests` → 1658 / 1658 pass
- `cargo clippy -p multitude --all-targets --all-features -- -D warnings` clean
- `cargo clippy -p multitude --all-targets --no-default-features -- -D warnings` clean
- `cargo fmt --check` clean

File: crates/multitude/src/arena/primitives.rs [hunk 3]
```diff
@@ -91,41 +91,50 @@ impl<A: Allocator + Clone> Arena<A> {
...
-            }
-
-            let needed = layout.size() + layout.align().saturating_sub(core::mem::align_of::<usize>());
-            self.refill_local(needed)?;
-        }
+        let needed = layout.size() + layout.align().saturating_sub(core::mem::align_of::<usize>());
+        self.refill_local(needed)?;
+        // `refill_local` post-condition: the refreshed chunk fits the
+        // request, so the retry cannot miss the bump-fit gate.
+        self.allocate_layout(layout)
```
[the remark this answers is not in the document]
Comment (geeknoid) [reply crates/multitude/src/arena/primitives.rs]: Fixed — converted `allocate_layout_slow` to a bounded refill-and-retry loop instead of tail-recursing into `allocate_layout`. Each iteration re-probes `try_bump_fit` after the refill, so a reentrant `Drop` that installs a too-small chunk simply triggers another refill rather than recursing back through the public entrypoint. Routes oversized requests once up front before entering the loop.

File: crates/multitude/src/arena/primitives.rs [hunk 4]
```diff
@@ -170,42 +179,40 @@ impl<A: Allocator + Clone> Arena<A> {
...
+        if layout.size() > self.provider.max_normal_alloc {
+            return self.allocate_shared_oversized_layout(layout);
         }
+        self.refill_shared(
+            layout
+                .size()
+                .saturating_add(layout.align().saturating_sub(core::mem::align_of::<usize>())),
+        )?;
+        // Refill post-condition: the refreshed chunk fits the request.
+        self.allocate_shared_layout(layout)
```
[the remark this answers is not in the document]
Comment (geeknoid) [reply crates/multitude/src/arena/primitives.rs]: Fixed — same pattern applied to `allocate_shared_layout_slow`: bounded refill-and-retry loop with per-iteration bump-fit re-probe, oversized routing handled once up front. The pre-existing slice slow paths in `inner_slice.rs` use the same tail-recursive shape (predating this PR), but that's out of scope here — this fix is consistent with the loop-based pattern already used in `impl_alloc_slice_local_with` and the shared-flavor sibling.

File: crates/multitude/src/arena/alloc_unsized.rs [hunk 5]
```diff
@@ -74,27 +74,19 @@ impl<A: Allocator + Clone> Arena<A> {
...
-            // were routed above), and `aligned_addr` lies inside a live
-            // chunk whose top is `<= isize::MAX as usize` on every real
-            // platform. The sum therefore fits in `usize` and cannot
-            // overflow.
+            // SAFETY: `aligned_addr` lies inside a live chunk whose
+            // bump extent is capped at `max_bump_extent <= MAX_CHUNK_BYTES`,
+            // and `bumped` is bounded by the caller's `Layout` invariant
+            // (`size <= isize::MAX`). The sum fits in `usize`.
             let end_addr = unsafe { aligned_addr.checked_add(bumped).unwrap_unchecked() };
             let new_drop_back_addr = drop_back_addr.saturating_sub(entry_size);
```
[the remark this answers is not in the document]
Comment (geeknoid) [reply crates/multitude/src/arena/alloc_unsized.rs:96]: Fixed — added `check_isize_overflow(layout.size(), layout.align())?` at the top of `try_reserve_dst_local_with_entry` to reject layouts where `size + align > isize::MAX` before any unchecked arithmetic. Updated the explanatory comment: the previous text claimed the alignment cap + oversized routing subsumed the check, but as you noted the bump-probe's `aligned_addr.checked_add(bumped).unwrap_unchecked()` runs *before* the oversized branch and would invoke UB on `None` for huge sizes.

File: crates/multitude/src/arena/alloc_unsized.rs [hunk 6]
```diff
@@ -188,25 +187,18 @@ impl<A: Allocator + Clone> Arena<A> {
...
```
[the remark this answers is not in the document]
Comment (geeknoid) [reply crates/multitude/src/arena/alloc_unsized.rs:207]: Fixed — same `check_isize_overflow` guard added at the top of `try_reserve_dst_shared_with_entry`, with a comment pointing at the local sibling for the rationale.

**Review by ralfbiedert — APPROVED** [review]

File: crates/multitude/src/arena/primitives.rs [hunk 7]
```diff
@@ -91,39 +91,65 @@ impl<A: Allocator + Clone> Arena<A> {
...
+        let align = layout.align().max(1);
+        let bumped = layout.size().max(1);
+
+        // Single bump-fit probe. The `max_normal_alloc` check and the
+        // refill loop are deferred to the cold slow path so the hot
+        // path avoids a 2-level pointer chase through `Arc<ChunkProvider>`.
+        let data_ptr = self.current_local.data_ptr.get();
+        let drop_back_ptr = self.current_local.drop_back.get();
+        let fit = try_bump_fit(data_ptr, drop_back_ptr, align, bumped, 0);
+        if fit.fits {
```
[the remark this answers is not in the document]
Comment (geeknoid) [reply crates/multitude/src/arena/primitives.rs:110]: Not an issue.
